materials_io/utils/interface.py

```diff
@@ -94,21 +124,55 @@ def get_adapter(name: str) -> BaseAdapter:
     return mgr.driver
 
 
-def execute_parser(name, group, context=None, adapter=None):
-    """Invoke a parser on a certain group of data
+def execute_parser(name, group, global_context=None,
+                   adapter_context: Union[None, dict] = None,
+                   parser_context: Union[None, dict] = None,
+                   adapter: Union[None, str] = None,
+                   adapter_map: Union[None, str] = None,
+                   default_adapter: Union[None, str] = None):
+    """Invoke a parser on a certain group of data AND apply appropriate adapter to metadata.
 
     Args:
         name (str): Name of the parser
         group ([str]): Paths to group of files to be parsed
-        context (dict): Context of the files, used in adapter and parser
+        global_context (dict): Context of the files, used for every parser and adapter
```

Comment:
I originally created the `execute_parser` function so that you can interact with a parser without having to know that other parsers exist. The idea is to keep the interface as simple as possible. Adding in `global context,` `adapter_map` and `default_adapter` violate that goal.

I'd propose creating a separate function for "run a parser given all of the information about how to run any of the parsers" if you find it necessary. This function is just as complicated as `run_all_parsers` and I'm not sure why it is needed.

Reply:
my goal is to just run that parser with all of the information I need about that parser. So what if I create a second function called 'parse_and_adapt()', that needs to do the following: 

1: Get the parser and adapter context for just 1 parser/adapter pair (rather than '@all')
-- *Side question: does this just look like 'name' or '@name'?*
2. get raw metadata by running execute_parser() with the parser context (no changes to execute_parser needed) 
3. Starting w/ outputted metadata from Step 2 (if it's not None/Error), then transform it using the adapter. 
4. Return transformed metadata.

Thoughts?

Reply:
That sounds like a good plan. I'm still a little bit concerned about a function which runs only a single parser yet requires the context/adapter map for all of them, but think you could get around them if clearly-documented.

What about making `parse_and_adapt` run all parsers. That way the options and the output are more consistent with each other. In a sense, it would be like the current `run_all_parsers` that only functions on a specified list of files rather than walking through all directories. If you wanted to only run a subset of the parsers, you could provide the OK list as an input (kind of like `run_all_parsers`).

Do you think that will give you what you need for XtractHub?

Reply:
If you think that's a good idea, feel free to rename `run_all_parsers` to `run_all_parsers_on_directory` or something like that so that it is clear how it is different than

Reply:
Thanks for your reply. I think the point of separating this into its own 'run a single parser' function is to fit the 'future proofing' I've done on Xtract's end that allows us to invoke single parsers rather than all of them on a group in order to avoid extraneous parser executions. Then I'd manually merge the records on the backend before generating a search index over them. 

That said, maybe I'm misunderstanding the point of the context, and whether I even need it if just running on one file. My thinking behind this PR is really "run_all_parsers() runs great and outputs the metadata I need because it can properly leverage the 'match' adapter. I want to do the same for a single group->parser mapping as well"

Reply:
> I want to do the same for a single group->parser mapping as well

This makes things really clear and is in line with what I was expecting you to do?

I think that the modifications I suggest to "parse_and_adapt" above will be able to suit that case. Is the difference that you want something that uses exactly one parser and what I propose can take >1?

Reply:
Ah okay, I think I understand now -- so in this world we can have 'run_all_parsers_on_directory' and 'run_all_parsers_on_group'.  The former is just run_all_parsers as we have it now, and the latter is 'given a group', then run all parsers. And if I want to run just one parser, I could denote that in the include_parsers argument.